va/va_prot.h

```diff
@@ -0,0 +1,515 @@
+  VA_TEE_EXECUTE_FUNCTION_ID_GET_FIRMWARE_VERSION = 0x00000002,
+
+} VA_TEE_EXECUTE_FUNCTION_ID;
+
+/** \brief Input/Output buffer of VAProtectedSessionExecuteBuffer */
+typedef struct _VAProtectedSessionBuffer {
+  /*
+   * This is used when this buffer refer to output buffer. The maximum size of
+   * data that the driver can return in the output buffer. It is not used for
+   * input buffer.
+   */
+  uint32_t max_data_size;
+  /*
+   * If it is used for input buffer, it is the size of the input data. If it is
+   * used for output buffer, it is the returns size of the output data written
+   * by the driver.
+   */
+  uint32_t data_size;
+  /*
+   * data pointer of this buffer
+   */
+  void *data;
+  uint32_t va_reserved[VA_PADDING_LOW];
+} VAProtectedSessionBuffer;
+
+/** \brief Buffer for vaProtectedSessionExecute() */
+typedef struct _VAProtectedSessionExecuteBuffer {
+  /** \brief Specify the function to execute. It is IHV's implementation
+   * specific */
+  uint32_t function_id;
+  /** \brief Input buffer */
+  VAProtectedSessionBuffer input;
+  /** \brief Output buffer */
+  VAProtectedSessionBuffer output;
+  /** \brief Return the result of this function. The status result is IHV's
+   * implementation specific */
+  uint32_t status;
+  uint32_t va_reserved[VA_PADDING_LOW];
+} VAProtectedSessionExecuteBuffer;
+
+/**
+ * \brief Create a protected session
+ *
+ * Create a protected session
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] config_id             configuration for the protected session
+ * @param[out] protected_session    created protected session id upon return
+ */
+VAStatus vaCreateProtectedSession(VADisplay dpy, VAConfigID config_id,
+                                  VAProtectedSessionID *protected_session);
+
+/**
+ * \brief Destroy a protected session
+ *
+ * Destroy a protected session
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] protected_session     protected session to be destroyed
+ */
+VAStatus vaDestroyProtectedSession(VADisplay dpy,
+                                   VAProtectedSessionID protected_session);
+
+/**
+ * \brief Attach a protected content session to VA context
+ *
+ * Attach a protected content session to the context to enable
+ * decryption/protection
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] id                    the VA decode/encode/vp context
+ * @param[in] protected_session     the protected session to attach
+ */
+VAStatus vaAttachProtectedSession(VADisplay dpy, VAGenericID id,
+                                  VAProtectedSessionID protected_session);
+
+/**
+ * \brief  Detach the protected content session from the VA context
+ *
+ * Detach protected content session of the context to disable
+ * decryption/protection
+ *
+ * @param[in] dpy                   the VA display
+ * @param[in] id                    TEE client id to be detached
+ */
+VAStatus vaDetachProtectedSession(VADisplay dpy, VAGenericID id);
+
```

Comment:
On Intel platform, a decode context could be attached to multiple protected sessions and visa-versa i.e. a protected session could be attached to multiple decode context.
For example,
1. A decode context could be attached to multiple protected sessions
Video bit-streaming is adaptive so a vendor could decide to apply multiple levels of protections across different resolutions of videos **for example** say video starts as 360p then improves to 480p then 720p then 1080p then 1440p then 4k. Vendor could say that video resolution <= 720p does not require HWDRM(L3 WV)  but 1080p to 1440p requires some HWDRM(L2 WV) and 4K requires full HWDRM(L1 WV). Then a media player will create 2 protected sessions with **different security requirements** and apply these protected sessions to this video stream. 

2. A protected session could be attached to multiple decode context.
Multiple video bit-streaming requires fetching multiple videos in a single session **for example** when a user is doing binge watching where current video is ending and next video stream has started in a window then most of time vendors use same license across multiple video streams to avoid interruption to user.

Reply:
if a decode context could be attached to multiple protected session, seems the parameters of vaDetachProtectedSession is not enough, for example, I attach two protect sessions to one decode context, then I want to detach one of them, how to set the parameter, seems vaDetachProtectedSession will detach all protected session from the decode context

Reply:
No Attach/Detach are per frame boundary. For example,
Attach(DecodeCtx, Session 1)
BeginFrame n
RenderFrame n
EndFrame n
Detach(Session 1)
...
Attach(DecodeCtx, Session 2)
BeginFrame n+x
RenderFrame n+x
EndFrame n+x
Detach(Session 2)
...
I believe Ryan has called this out in comments.

Reply:
so, application have to attach and detach perframe,  "attach one time, process multiple frame" is not allowed, right? even there are no change with the protect session.

Reply:
Technically, Yes but if application does not call detach then protected session remains attach for full decode duration. 
Here is WV productized scenario and solution - certain bit-streams in HWDRM has combinations of encrypted and clear frames so question was how to use attach/detach. 
Let us say frame 1 is encrypted but frame 2 is clear then technically app could do as I mentioned above but this arrangement although very generic but puts extra work on app so WA is always attach protected session and for clear frame mention encrypted bytes a 0.
Attach(DecodeCtx, Session 1)
BeginFrame n
RenderFrame n (encrypted bytes x, clear bytes y)
EndFrame n
Detach(Session 1)
...
BeginFrame n+x
RenderFrame n+x (encrypted bytes 0, clear bytes x)
EndFrame n+x